src/QsCompiler/QsTargeting/CapabilityLeveler.fs

```diff
@@ -0,0 +1,243 @@
+        base.onValueUpdate(s)
+
+    override this.onVariableDeclaration(s) =
+        if s.Kind = QsBindingKind.MutableBinding 
+        then holder.LocalLevel <- CapabilityLevel.Advanced
+        else
+            if not (isResult s.Rhs)
+            then holder.LocalLevel <- CapabilityLevel.Advanced
+        base.onVariableDeclaration(s)
+
+/// Walker for setting capability levels based on scopes.
+and private ScopeLeveler(holder : CapabilityInfoHolder) as this =
+    inherit ScopeWalker()
+
+    let kindXformer = new StatementLeveler(this, holder)
+
+    override this.StatementKind = upcast kindXformer
+
+    member this.Holder with get() = holder
+
+/// This syntax tree transformer fills in the CapabilityLevel fields in specializations,
+/// based on information gathered by the associated scope and other walkers.
+type TreeLeveler() =
+    inherit SyntaxTreeTransformation()
+
+    // Checks to see if a callable or specialization has a developer-specified capability level,
+    // which overrides the level computed from the code if it is present.
+    let checkForLevelAttributes (attrs : QsDeclarationAttribute seq) =
+        let isLevelAttribute (a : QsDeclarationAttribute) = 
+            match a.TypeId with
+            | Value udt -> udt.Namespace = BuiltIn.LevelAttribute.Namespace && udt.Name = BuiltIn.LevelAttribute.Name
+            | Null -> false
+        let getLevelFromArgument (a : QsDeclarationAttribute) =
+            match a.Argument.Expression with
+            | IntLiteral n -> let level = enum<CapabilityLevel> (int n)
+                              Some level
+            | ValueTuple args -> if args.Length = 1
+                                 then
+                                    match args.[0].Expression with
+                                    | IntLiteral n -> let level = enum<CapabilityLevel> (int n)
+                                                      Some level
+                                    | _ -> None
+                                 else None
+            | _ -> None
+        let levels = attrs |> Seq.filter isLevelAttribute |> List.ofSeq
+        match levels with
+        | [ level ] -> getLevelFromArgument level
+        | [] -> None
+        | _ -> None
+
+    let mutable currentOperationLevel = None : CapabilityLevel option
+
+    override this.Scope with get() = new ScopeTransformation()
+
+    override this.beforeCallable(c) =
+        currentOperationLevel <- c.Attributes |> checkForLevelAttributes
+        let result = base.beforeCallable(c)
+        result
+
+    override this.onSpecializationImplementation(s) =
+        let codeLevel = match s.Implementation with
+                        | Intrinsic -> CapabilityLevel.Minimal
+                        | External -> CapabilityLevel.Unset
+                        | Generated _ -> CapabilityLevel.Unset
+    // TODO: For generated specializations, we need to find the appropriate "body" declaration
```

Comment:
You know, this specifically makes me wonder if we might instead want to track it as inferred information. Let's maybe quickly discuss that option.